### PR TITLE
feat: support user-defined types in YAML extensions

### PR DESCRIPTION
**Pull request by kadinrabo**
Support parsing `UserDefinedType` (denoted by u!) in extension yaml

### CODE REVIEW

**Review by vbarua — APPROVED** [review]
Looks reasonable to me, but I'm not super familiar with the Go side of things.

cc: @zeroshade, @jacques-n: should be an easy review.

File: extensions/simple_extension_test.go
Comment (zeroshade) [inline extensions/simple_extension_test.go:72]: Can we also test that the type of the argument is correctly produced as a `*UserDefinedType`?

File: types/parser/type_parser.go
Comment (zeroshade) [inline types/parser/type_parser.go:125]: `RetType()` probably needs to be updated to handle a user-defined type scenario. I'm pretty sure `SimpleTypeNameToType` will not currently return a user-defined type, it'll error out.

**Review by zeroshade — CHANGES_REQUESTED** [review]

**Comment by jacques-n** [comment]
I believe @anshuldata is or will be working on a patch to replace the hand parsing with antlr grammar. That should address several gaps in the current function parsing. I'd be inclined to wait until that is merged before trying to solve this gap (as some of it may come for cheap/free).

**Comment by vbarua** [comment]
@jacques-n do you think that will be available next week? We have some stuff internally that's blocked on us being able to handle user-defined types. Aside from potentially redundant work on our end, would there be any harm in updating the hand parser.

File: extensions/simple_extension_test.go
Comment (kadinrabo) [reply extensions/simple_extension_test.go:91]: I added a nullability check too 👍

**Comment by zeroshade** [comment]
@vbarua It's unlikely that something as large as replacing with an antlr grammar will be merged next week as I'm sure that would need significant review and testing. My personal opinion would be that it makes sense to update this if you need it. I doubt there's any harm besides the redundant work.

@jacques-n thoughts?

**Comment by jacques-n** [comment]
I think this patch is building debt on debt. 

In general, I find it to be a bad habit to accept something we know we're going to delete in a couple of weeks (unless we're fixing a critical regression). It sets a bad precedent for new contributors on how we work. If we know we're going in a certain direction and one contributor wants to go there faster than another, let's just have them take us to the target outcome.

I'm -0 on this patch. I'm not going to block.

**Review by zeroshade — APPROVED** [review]
LGTM assuming that @jacques-n has no objections

**Comment by vbarua** [comment]
Based on 
> I'm -0 on this patch. I'm not going to block.

I'm going to merge this in.

**Pull request merged**
